### DISCUSSION TITLE
How to allocate minimum heap with picolibc linker?

### DISCUSSION THREAD

**Discussion by ivanmak0509** [picolibc/picolibc#537] [Q&A] [answered]
How can I allocate a minimum heap without modifying the default linker? The main reason is to avoid situations where I use up all of the ram and have little to no heap remaining *without* getting any error at link time. Is there an intended way to do this?

**Comment by keith-packard** [comment]
Because the default linker script places the stack above the heap, you be able to set the stack size to a value large enough for both your required stack *and* minimum heap and the linker will make sure there's space for both. Here's an example for a device with 32kB flash and 4kB of ram which reserves 3kB for heap and stack.

```
__flash =      0x00000000;
__flash_size = 0x00008000;
__ram =        0x20000000;
__ram_size   = 0x00001000;
__stack_size = 3k;

INCLUDE picolibc.ld
```

**Reply by ivanmak0509** [reply]
aha, pardon the late reply, but why is line 250 in picolibc.ld.in then setting `__heap_end ` to the begginning of the stack? Or is some of the stack region also allowed to be used as heap?

**Reply by keith-packard** [reply, marked as answer]
oh, you're so right -- I'd forgotten that there was a check to make sure the heap didn't smash into the defined stack space.

How about a simple addition to the linker script that would let you specify a minimum heap size while still allowing the heap to grow up to the bottom of the reserved stack area?

```
diff --git a/picolibc.ld.in b/picolibc.ld.in
index 44900fed1..ec0c79a00 100644
--- a/picolibc.ld.in
+++ b/picolibc.ld.in
@@ -247,6 +247,12 @@ SECTIONS
 
        /* Make the rest of memory available for heap storage */
        PROVIDE (__heap_start = __end);
+
+       /* Allow a minimum heap size to be specified */
+       .heap (NOLOAD) : {
+               . += (DEFINED(__heap_size_min) ? __heap_size_min : 0);
+       } >ram :ram
+
        PROVIDE (__heap_end = __stack - (DEFINED(__stack_size) ? __stack_size : @DEFAULT_STACK_SIZE@));
        PROVIDE (__heap_size = __heap_end - __heap_start);
 
```

With this, you'd add `__heap_size_min = 4k` to your linker script fragment and the linker would ensure there was enough space.

**Reply by ivanmak0509** [reply]
cool, Ill copy the oiginal linker into my project dir and modify it from there, thanks.

**Comment by keith-packard** [comment]
If this works, I'm happy to fold it into the repository; it's harmless if unused.

**Reply by ivanmak0509** [reply]
Sure! It seams to be working fine, it reserves the extra space and throws an error when the ram overflows.

**Comment by keith-packard** [comment]
Merged to main, will be in 1.8.4
